snuba/query/matchers.py

```diff
@@ -0,0 +1,317 @@
+    This is meant to represent what a group is in a regex.
+
+    Params can be nested, and they can wrap any valid Pattern.
+    """
+
+    name: str
+    pattern: Pattern[TMatchedType]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        result = self.pattern.match(node)
+        if not result:
+            return None
+        return result.merge(MatchResult({self.name: node}))
+
+
+@dataclass(frozen=True)
+class AnyExpression(Pattern[Expression]):
+    """
+    Matches any expression of the type provided.
+    This allows us to match any Expression since the Any class cannot
+    match abstract classes (like Expression)
+    """
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if isinstance(node, Expression) else None
+
+
+@dataclass(frozen=True)
+class Any(Pattern[TMatchedType]):
+    """
+    Match any concrete expression/scalar of the type provided.
+    """
+
+    type: Type[TMatchedType]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if isinstance(node, self.type) else None
+
+
+@dataclass(frozen=True)
+class String(Pattern[str]):
+    """
+    Matches one specific string.
+    """
+
+    value: str
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class OptionalString(Pattern[Optional[str]]):
+    """
+    Matches one specific string (or None).
+    """
+
+    value: Optional[str]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class AnyOptionalString(Pattern[Optional[str]]):
+    """
+    Matches any string including the None value. This cannot be done with
+    Any(type) because that cannot match Union[str, None].
+    """
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node is None or isinstance(node, str) else None
+
+
+@dataclass(frozen=True)
+class Or(Pattern[TMatchedType]):
+    """
+    Union of multiple patterns. Matches if at least one is a valid match
+    and returns the first valid one.
+    """
+
+    patterns: Sequence[Pattern[TMatchedType]]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        for p in self.patterns:
+            ret = p.match(node)
+            if ret:
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[ColumnExpr]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each, a Pattern can be
+    provided. If one of these Patterns is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[str]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
```

Comment:
Note that the ordering here should be kept in sync with #963, whichever one of these lands fist.

Reply:
that makes sense. This will land first but I will make the change here before it lands since 963 is already accepted and will land tomorrow morning.